app/models/mixins/aggregation_mixin.rb

```diff
@@ -1,23 +1,22 @@
 module AggregationMixin
   extend ActiveSupport::Concern
   included do
-    virtual_column :aggregate_cpu_speed,       :type => :integer, :uses => :hosts
-    virtual_column :aggregate_cpu_total_cores, :type => :integer, :uses => :hosts
-    virtual_column :aggregate_physical_cpus,   :type => :integer, :uses => :hosts
-    virtual_column :aggregate_memory,          :type => :integer, :uses => :hosts
-    virtual_column :aggregate_vm_cpus,         :type => :integer, :uses => :vms_and_templates
-    virtual_column :aggregate_vm_memory,       :type => :integer, :uses => :vms_and_templates
-    virtual_column :aggregate_disk_capacity,   :type => :integer, :uses => :hosts
+    virtual_aggregate :aggregate_vm_memory, :vm_hardwares, :sum, :memory_mb
+    virtual_aggregate :aggregate_vm_cpus, :vm_hardwares, :sum, :cpu_sockets
+    virtual_aggregate :aggregate_physical_cpus, :host_hardwares, :sum, :cpu_sockets
+    virtual_aggregate :aggregate_disk_capacity, :host_hardwares, :sum, :disk_capacity
+    virtual_aggregate :aggregate_memory, :host_hardwares, :sum, :memory_mb
+    virtual_aggregate :aggregate_cpu_total_cores, :host_hardwares, :sum, :cpu_total_cores
+    virtual_aggregate :aggregate_cpu_speed, :host_hardwares, :sum, :aggregate_cpu_speed
 
-    alias_method :all_hosts,              :hosts
-    alias_method :all_host_ids,           :host_ids
-    alias_method :all_vms_and_templates,  :vms_and_templates
-    alias_method :all_vm_or_template_ids, :vm_or_template_ids
-    alias_method :all_vms,                :vms
-    alias_method :all_vm_ids,             :vm_ids
-    alias_method :all_miq_templates,      :miq_templates
-    alias_method :all_miq_template_ids,   :miq_template_ids
+    def aggregate_hardware(from, field, targets = nil)
+      from      = from.to_s.singularize
+      select    = field == :aggregate_cpu_speed ? "cpu_total_cores, cpu_speed" : field
+      targets ||= send("all_#{from}_ids")
+      targets   = targets.collect(&:id) unless targets.first.kind_of?(Integer)
+      hdws      = Hardware.where("#{from}_id".to_sym => targets).select(select)
 
-    include AggregationMixin::Methods
+      hdws.inject(0) { |t, hdw| t + hdw.send(field).to_i }
```

Comment:
Seems like this entire method can be trimmed down to just this:

```ruby
def aggregate_hardware(from, field, targets = nil)
  from      = from.to_s.singularize
  targets ||= send("all_#{from}_ids")
  targets   = targets.collect(&:id) unless targets.first.kind_of?(Integer)

  Hardware.where("#{from}_id".to_sym => targets).sum(field)
end
```

Also, we probably could drop the `collect(&:id)` as this does seem to work:

```ruby
# aggregate_hardware(:computer_systems, :memory_mb, targets)
Hardware.where(:computer_systems_id => ContainerProject.first.computer_systems).sum(:memory_mb)
```

Reply:
I would like to agree. I would. Very much. But I don't. This one makes me nervous. And `spec/models/metric/statistic_spec.rb:49` is really unhappy with this.

Reply:
I get defaulting the field to computer systems
also, I feel more comfortable using `:computer_systems` than the id conterpart

I'm just a little confused with the container project.first call. not quite groking the first.

Reply:
> I'm just a little confused with the container project.first call. not quite groking the first.

Oh, sorry, that example was from my IRB session when I was testing this with some data.

My point was that we could drop the `targets = targets.collect(&:id) ...` line since relations seem to be allowed as a value in a `.where` that would normally accept an Array of ids.

Reply:
> My point was that we could drop the targets = targets.collect(&:id) ...

hmm. @d-m-u - I thought we already said this before. ActiveRecord calls `to_param` on objects, which is essentially `id` (unless it was overridden - which we do not)